internal/receiver/discoveryreceiver/statussources/statements.go

```diff
@@ -0,0 +1,160 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statussources
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+)
+
+var (
+	ReceiverCreatorRegexp = regexp.MustCompile(`receiver_creator/`)
+	receiverNameRegexp    = regexp.MustCompile(`^(?P<type>[^/]+)/(?P<name>.*)$`)
+	EndpointTargetRegexp  = regexp.MustCompile(`{endpoint=[^}]*}/`)
+	endpointIDRegexp      = regexp.MustCompile(`^.*{endpoint=.*}/(?P<id>.*)$`)
+	undesiredFields       = []string{"ts", "msg", "level"}
+)
+
+// Statement models a zapcore.Entry but defined here for usability/maintainability
+type Statement struct {
+	Message    string
+	Fields     map[string]any
+	Level      string
+	Time       time.Time
+	LoggerName string
+	Caller     zapcore.EntryCaller
+	Stack      string
+}
+
+func NewZapCoreEncoder() zapcore.Encoder {
+	return zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig())
+}
+
+// StatementFromZapCoreEntry converts the Entry to a Statement using the provided encoder, which is assumed to be
+// a JSONEncoder (unexpected from zapcore) to obtain the target Fields.
+func StatementFromZapCoreEntry(encoder zapcore.Encoder, entry zapcore.Entry, fields []zapcore.Field) (*Statement, error) {
+	statement := &Statement{
+		Message:    entry.Message,
+		Level:      entry.Level.String(),
+		Time:       entry.Time,
+		LoggerName: entry.LoggerName,
+		Caller:     entry.Caller,
+		Stack:      entry.Stack,
+	}
+	var err error
+	var entryBuffer *buffer.Buffer
+
+	if entryBuffer, err = encoder.EncodeEntry(entry, fields); err != nil {
+		return nil, fmt.Errorf("failed encoding zapcore.Entry: %w", err)
+	}
+
+	b := entryBuffer.Bytes()
+	if err = json.Unmarshal(b, &statement.Fields); err != nil {
+		return nil, fmt.Errorf("failed representing encoded zapcore.Entry (%s) as json: %w", b, err)
+	}
+
+	for _, undesiredField := range undesiredFields {
+		delete(statement.Fields, undesiredField)
+	}
+
+	return statement, nil
+}
+
+func (s *Statement) ToLogRecord() plog.LogRecord {
+	logRecord := plog.NewLogRecord()
+	if s == nil {
+		return logRecord
+	}
+
+	logRecord.SetTimestamp(pcommon.NewTimestampFromTime(s.Time))
+	logRecord.SetObservedTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+	logRecord.Body().SetStringVal(s.Message)
+	logRecord.SetSeverityText(s.Level)
+	fieldMap := pcommon.NewMapFromRaw(s.Fields)
+	fieldMap.CopyTo(logRecord.Attributes())
```

Comment:
I think you can do `logRecord.Attributes().fromRaw(s.fields)`

Reply:
updated w/ new option after rebasing.